File: services/scoringService.ts
import { Question, QuestionType, SubjectId, TestResult, UserAnswers } from '../types';

export function scoreQuestion(question: Question, selectedIds: string[]): { score: number; max: number; correct: boolean } {
  const correctIds = question.correctOptionIds || [];

  if (question.type === QuestionType.MULTIPLE) {
    const max = 2;
    if (selectedIds.length === 0) return { score: 0, max, correct: false };

    const correctSet = new Set(correctIds);
    const selectedSet = new Set(selectedIds);
    const missed = correctIds.filter(id => !selectedSet.has(id)).length;
    const extra = selectedIds.filter(id => !correctSet.has(id)).length;
    const mistakes = missed + extra;

    if (mistakes === 0) return { score: 2, max, correct: true };
    if (mistakes === 1) return { score: 1, max, correct: false };
    return { score: 0, max, correct: false };
  }

  const isCorrect = selectedIds.length === 1 && correctIds.includes(selectedIds[0]);
  return { score: isCorrect ? 1 : 0, max: 1, correct: isCorrect };
}

export function calculateTestResult(questions: Question[], answers: UserAnswers): TestResult {
  const subjectScores: TestResult['subjectScores'] = {
    [SubjectId.ENGLISH]: { score: 0, max: 0 },
    [SubjectId.TGO]: { score: 0, max: 0 },
    [SubjectId.ALGO]: { score: 0, max: 0 },
    [SubjectId.DB]: { score: 0, max: 0 },
    [SubjectId.M001_PEDAGOGIKA]: { score: 0, max: 0 },
    [SubjectId.M001_PSYCHOLOGY]: { score: 0, max: 0 },
    [SubjectId.M002_PEDAGOGIKA]: { score: 0, max: 0 },
    [SubjectId.M002_SPEECH_DEV]: { score: 0, max: 0 },
  };

  let totalScore = 0;
  let maxScore = 0;
  let correctCount = 0;

  questions.forEach(question => {
    const result = scoreQuestion(question, answers[question.id] || []);
    totalScore += result.score;
    maxScore += result.max;
    if (result.correct) correctCount += 1;

    if (subjectScores[question.subjectId]) {
      subjectScores[question.subjectId].score += result.score;
      subjectScores[question.subjectId].max += result.max;
    }
  });

  return {
    totalScore,
    maxScore,
    subjectScores,
    correctCount,
    totalQuestions: questions.length,
  };
}

File: ios/MagisCore/Services/ScoringService.swift
import Foundation

enum ScoringService {
    static func calculateResult(questions: [Question], answers: UserAnswers, userName: String = "Пайдаланушы") -> TestResult {
        var totalScore = 0
        var maxScore = 0
        var correctCount = 0
        var subjectScores: [SubjectId: SubjectScore] = [:]

        let grouped = Dictionary(grouping: questions, by: { $0.subjectId })

        for (subjectId, subjectQuestions) in grouped {
            var subjectScore = 0
            var subjectMax = 0

            for question in subjectQuestions {
                let qMax = question.type == .multiple ? 2 : 1
                subjectMax += qMax

                let userAnswer = answers[question.id] ?? []
                let correct = Set(question.correctOptionIds)
                let selected = Set(userAnswer)

                if question.type == .multiple {
                    if selected == correct {
                        subjectScore += 2
                    } else if selected.isSubset(of: correct) && !selected.isEmpty {
                        subjectScore += 1
                    }
                } else {
                    if selected == correct {
                        subjectScore += 1
                    }
                }
            }

            totalScore += subjectScore
            maxScore += subjectMax
            if subjectScore == subjectMax { correctCount += subjectQuestions.count }
            else {
                let fullyCorrect = subjectQuestions.filter { q in
                    let userAnswer = answers[q.id] ?? []
                    return Set(userAnswer) == Set(q.correctOptionIds)
                }.count
                correctCount += fullyCorrect
            }

            subjectScores[subjectId] = SubjectScore(score: subjectScore, max: subjectMax)
        }

        let wrongIds = questions
            .filter { q in
                let userAnswer = answers[q.id] ?? []
                return Set(userAnswer) != Set(q.correctOptionIds)
            }
            .map(\.id)

        return TestResult(
            totalScore: totalScore,
            maxScore: maxScore,
            subjectScores: subjectScores,
            correctCount: correctCount,
            totalQuestions: questions.count,
            userName: userName,
            endDate: Date(),
            wrongQuestionIds: wrongIds
        )
    }

    static func isPassing(_ result: TestResult) -> Bool {
        Double(result.totalScore) / Double(result.maxScore) >= Constants.passingThreshold
    }

    static func percentage(_ result: TestResult) -> Double {
        guard result.maxScore > 0 else { return 0 }
        return Double(result.totalScore) / Double(result.maxScore) * 100
    }

    static func getWeakSubjects(_ result: TestResult) -> [(SubjectId, Double)] {
        result.subjectScores
            .map { (id, score) in
                let pct = score.max > 0 ? Double(score.score) / Double(score.max) : 0
                return (id, pct)
            }
            .filter { $0.1 < 0.5 }
            .sorted { $0.1 < $1.1 }
    }
}
